src/main/java/org/dice_group/grp/grammar/digram/Digram.java

```diff
@@ -82,4 +78,51 @@ public void setNoOfOccurences(long occurences) {
 		this.occurences = occurences;
 	}
 	
+
+
+	@Override
+	public int hashCode() {
+		final int prime = 31;
+		int result = 1;
+		result = prime * result + ((edgeLabel1 == null) ? 0 : edgeLabel1.hashCode());
+		result = prime * result + ((edgeLabel2 == null) ? 0 : edgeLabel2.hashCode());
+		result = prime * result + ((external == null) ? 0 : external.hashCode());
+		return result;
+	}
+
+	@Override
+	public boolean equals(Object obj) {
```

Comment:
all the edgeLabelX == null can be removed, if an edgeLabel is null there is something wrong. :)

Reply:
Again, you could also throw an exception. Or use [`Objects.requireNonNull()`](https://docs.oracle.com/javase/7/docs/api/java/util/Objects.html#requireNonNull(T,%20java.lang.String))